File: backend/data/locations.py
"""Location registry backed by shared/locations.json (also used by the frontend)."""

import json
from functools import lru_cache
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from core.errors import InvalidLocationError, NoMonitoringDataError
from core.geo import haversine_km

SHARED_LOCATIONS = Path(__file__).resolve().parents[2] / "shared" / "locations.json"

Location = Dict[str, Any]


def _normalise(name: str) -> str:
    return " ".join(name.strip().lower().replace("-", " ").replace("_", " ").split())


@lru_cache(maxsize=1)
def _registry() -> Dict[str, Any]:
    return json.loads(SHARED_LOCATIONS.read_text(encoding="utf-8"))


def list_locations() -> List[Location]:
    return list(_registry()["locations"])


def get_location(name: str) -> Location:
    """Resolve a user-supplied location name, raising a typed error when unsupported."""
    key = _normalise(name or "")
    if not key:
        raise InvalidLocationError("Please choose a location to analyze.")

    for location in _registry()["locations"]:
        names = [location["id"], location["name"], *location.get("aliases", [])]
        if key in {_normalise(n) for n in names}:
            return location

    for uncovered in _registry()["noCoverage"]:
        if key == _normalise(uncovered):
            raise NoMonitoringDataError(
                f"No monitoring stations are deployed in {uncovered} yet."
            )

    raise InvalidLocationError(f"'{name.strip()}' is not a recognised monitoring location.")


def find_profile(profile_id: str) -> Optional[Location]:
    return next((loc for loc in _registry()["locations"] if loc["id"] == profile_id), None)


def nearest_profile(lat: float, lon: float) -> Tuple[Location, float]:
    """Closest demo profile to a coordinate (used only by demo providers and demo history)."""
    distances = [(loc, haversine_km(lat, lon, loc["lat"], loc["lon"])) for loc in _registry()["locations"]]
    return min(distances, key=lambda item: item[1])

File: frontend/src/data/locations.ts
import registry from '@shared/locations.json';
import { AppError } from '../services/errors';
import type { MapStation, WaterBody } from '../types/environment';

/** Location profile shared with the backend (shared/locations.json). */
export interface LocationProfile {
  id: string;
  name: string;
  region: string;
  lat: number;
  lon: number;
  aliases: string[];
  air: { stationId: string; stationName: string; pm25: number | null; pm10: number | null; no2: number | null; o3: number | null };
  water: {
    sensorId: string;
    sensorName: string;
    status: 'online' | 'degraded' | 'offline';
    ph: number | null;
    turbidity: number | null;
    temperature: number | null;
  };
  waste: { cameraId: string; cameraName: string; plastic: number; paper: number; other: number };
  trend: { air: number; water: number; waste: number };
  map: { stations: MapStation[]; waterBodies: WaterBody[] };
}

const data = registry as unknown as { locations: LocationProfile[]; noCoverage: string[] };

export const LOCATIONS: LocationProfile[] = data.locations;
export const NO_COVERAGE: string[] = data.noCoverage;

const normalise = (name: string) => name.trim().toLowerCase().replace(/[-_]/g, ' ').split(/\s+/).join(' ');

export function findLocation(name: string): LocationProfile | undefined {
  const key = normalise(name);
  return LOCATIONS.find((location) =>
    [location.id, location.name, ...location.aliases].some((candidate) => normalise(candidate) === key),
  );
}

/** Resolve a user-entered location or throw a user-friendly AppError. */
export function resolveLocation(name: string): LocationProfile {
  const key = normalise(name ?? '');
  if (!key) throw new AppError('INVALID_LOCATION', 'Please choose a location to analyze.');

  const location = findLocation(name);
  if (location) return location;

  const uncovered = NO_COVERAGE.find((candidate) => normalise(candidate) === key);
  if (uncovered) {
    throw new AppError('NO_MONITORING_DATA', `No monitoring stations are deployed in ${uncovered} yet.`);
  }
  throw new AppError('INVALID_LOCATION', `'${name.trim()}' is not a recognised monitoring location.`);
}

/**
 * The preset monitoring area closest to a coordinate pair.
 *
 * Used only by Demo Mode, which has fixtures for the preset areas and nothing else. When the
 * user has picked a real point, Demo Mode renders the nearest preset's fixtures — labelled as
 * demo data throughout the UI, so this is never presented as a reading from their location.
 */
export function nearestLocation(latitude: number, longitude: number): LocationProfile {
  // Equirectangular approximation: adequate for ranking a handful of cities, and it avoids
  // pulling in a full haversine for a demo-only fallback.
  const distanceSq = (location: LocationProfile) => {
    const dLat = location.lat - latitude;
    const dLon = (location.lon - longitude) * Math.cos((latitude * Math.PI) / 180);
    return dLat * dLat + dLon * dLon;
  };
  return LOCATIONS.reduce((closest, location) => (distanceSq(location) < distanceSq(closest) ? location : closest));
}
